### PR TITLE
[ROX-18762] : Update run state enum

### PR DESCRIPTION
**Pull request by charmik-redhat**
## Description

- Updated run state enum to replace `SUCCESS` state with `GENERATED` and `DELIVERED` states.
- In report history response, snapshots will have `isDownloadAvailable = true` if snapshot if notification method is `DOWNLOAD` and report exists in blob store.
- Updated report history pruning to match `GENERATED` and `DELIVERED` states' new enum numbers.

## Checklist
- [ ] Investigated and inspected CI test results
-  ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Manual testing
Unit tests

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.

### CODE REVIEW

**Review by md2119 — CHANGES_REQUESTED** [review]
Please update the pruner and send a re-review request

File: central/reports/common/searcher.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	}
+}
+
+// Replace match field queries of type Report State == SUCCESS and Report State == "SUCCESS" (Exact match)
+// Note : This func does no replacement
+//  1. if the query is a prefix query designed to match any non-complete prefix of 'SUCCESS'
+//  2. if the query is a regex query designed to match 'SUCCESS' among other matches
+//  3. if the query is a negation query designed to not match 'SUCCESS' or any of its prefixes
+//
+// To avoid these discrepancies we should stop supporting 'SUCCESS' as a value for "Report State" field in the long run.
```
Comment (md2119) [inline central/reports/common/searcher.go:50]: @charmik-redhat, please add a backend and frontend task to stop the use of SUCCESS in the API and UI.   We did not want to do it tech preview due to time constraints but we would like us to tackle this problem in GA.
BE: https://issues.redhat.com/browse/ROX-18979
FE: https://issues.redhat.com/browse/ROX-18980

File: central/reports/service/v2/convert.go [hunk 2]
```diff
@@ -376,38 +380,64 @@ func ConvertProtoNotifierSnapshotToV2(notifierSnapshot *storage.NotifierSnapshot
...
 
 // convertPrototoV2ReportSnapshot converts storage.ReportSnapshot to apiV2.ReportSnapshot
-func convertProtoReportSnapshotstoV2(snapshots []*storage.ReportSnapshot) []*apiV2.ReportSnapshot {
+func (s *serviceImpl) convertProtoReportSnapshotstoV2(snapshots []*storage.ReportSnapshot) ([]*apiV2.ReportSnapshot, error) {
 	if snapshots == nil {
-		return nil
+		return nil, nil
 	}
-	res := make([]*apiV2.ReportSnapshot, 0, len(snapshots))
+	blobNames, err := s.getExistingBlobNames(snapshots)
```
Comment (md2119) [inline central/reports/service/v2/convert.go:387]: This seems expensive. Why not search?

File: central/reports/service/v2/convert.go [hunk 3]
```diff
@@ -376,38 +380,64 @@ func ConvertProtoNotifierSnapshotToV2(notifierSnapshot *storage.NotifierSnapshot
...
+	blobs, err := s.blobStore.GetManyMetadata(allAccessCtx, blobNames)
+	if err != nil {
+		return nil, err
+	}
+
+	existingBlobNames := set.NewStringSet()
+	for _, b := range blobs {
+		existingBlobNames.Add(b.GetName())
+	}
+	return existingBlobNames, nil
```
[the remark this answers is not in the document]
Comment (charmik-redhat) [reply central/reports/service/v2/convert.go:442]: Do you mean search blob store with a query like `Blob name : name1 OR name2 OR ....`? I think `s.blobStore.GetManyMetadata(...)` would do a similar thing?

File: central/reports/service/v2/convert.go (hunk 2 above)
Comment (charmik-redhat) [reply central/reports/service/v2/convert.go:387]: Same question as [here](https://github.com/stackrox/stackrox/pull/7410#discussion_r1293838183)

**Comment by md2119** [comment]
Report history can only be obtained for one specific report configuration. Therefore, all snapshots prepared for the API response belong to one configuration. The blobname pattern is `configID/reportID`, therefore, fetching full blob metadata objects to decide if a blob for a specific configuration and report ID exists seems overkill.

Separately, the `GENERATED` state should be sufficient to determine whether the download is available. Otherwise, the state management becomes moot if we need to check other data points for making some decisions.

**Comment by charmik-redhat** [comment]
> Report history can only be obtained for one specific report configuration. Therefore, all snapshots prepared for the API response belong to one configuration. The blobname pattern is `configID/reportID`, therefore, fetching full blob metadata objects to decide if a blob for a specific configuration and report ID exists seems overkill.

Yeah I agree, fetching entire blob objects is an overkill. I think just fetching the IDs using `Search` on blob datastore should suffice for that.

> Separately, the `GENERATED` state should be sufficient to determine whether the download is available. Otherwise, the state management becomes moot if we need to check other data points for making some decisions.

We allow pruning blobs before it has been downloaded, because otherwise the non-downloaded ones can take up storage. So I think checking whether a blob exists or not would be important to determine if download is available or not.

**Comment by md2119** [comment]
> We allow pruning blobs before it has been downloaded, because otherwise the non-downloaded ones can take up storage. So I think checking whether a blob exists or not would be important to determine if download is available or not.

Not for GA: We should look at the possibility of adding another state. With `status = GENERATED/SUCCESS` and `isDownloadAvailable = false`, the question the consumer will have is why is the download file not available if the status indicates success.

**Pull request merged**